src/main/java/org/opensearch/ad/task/ADBatchTaskRunner.java

```diff
@@ -177,58 +197,395 @@ public ADBatchTaskRunner(
+        boolean isHCDetector = adTask.getDetector().isMultientityDetector();
+        if (isHCDetector && !adTaskCacheManager.topEntityInited(adTask.getDetectorId())) {
+            // Initialize top entities for HC detector
+            adTaskCacheManager.add(adTask);
+            threadPool.executor(AD_BATCH_TASK_THREAD_POOL_NAME).execute(() -> {
+                ActionListener<ADBatchAnomalyResultResponse> hcDelegatedListener = getInternalHCDelegatedListener(adTask);
+                ActionListener<String> internalHCListener = internalHCListener(adTask, transportService, hcDelegatedListener);
+                try {
+                    getTopEntities(adTask, internalHCListener);
+                } catch (Exception e) {
+                    internalHCListener.onFailure(e);
+                }
+            });
+            listener.onResponse(new ADBatchAnomalyResultResponse(clusterService.localNode().getId(), false));
+        } else {
+            // Execute AD task for single entity detector or HC detector which top entities initialized
+            forwardOrExecuteADTask(adTask, transportService, listener);
+        }
+    }
 
-        ActionListener<ADBatchAnomalyResultResponse> delegatedListener = ActionListener.wrap(r -> { listener.onResponse(r); }, e -> {
-            listener.onFailure(e);
-            handleException(adTask, e);
+    private ActionListener<ADBatchAnomalyResultResponse> getInternalHCDelegatedListener(ADTask adTask) {
+        return ActionListener
+            .wrap(
+                r -> logger.debug("[InternalHCDelegatedListener]: running task {} on nodeId {}", adTask.getTaskId(), r.getNodeId()),
+                e -> logger.error("[InternalHCDelegatedListener]: failed to run task", e)
+            );
+    }
+
+    /**
+     * Create internal action listener for HC task. The action listener will be used in
+     * {@link ADBatchTaskRunner#getTopEntities(ADTask, ActionListener)}. Will call
+     * listener's onResponse method when get top entities done.
+     *
+     * @param adTask AD task
+     * @param transportService transport service
+     * @param listener action listener
+     * @return action listener
+     */
+    private ActionListener<String> internalHCListener(
+        ADTask adTask,
+        TransportService transportService,
+        ActionListener<ADBatchAnomalyResultResponse> listener
+    ) {
+        ActionListener<String> actionListener = ActionListener.wrap(response -> {
+            adTaskCacheManager.setTopEntityInited(adTask.getDetectorId());
+            int totalEntities = adTaskCacheManager.getPendingEntityCount(adTask.getDetectorId());
+            logger.info("total top entities: {}", totalEntities);
+            int numberOfEligibleDataNodes = nodeFilter.getNumberOfEligibleDataNodes();
+            int maxRunningEntities = Math
+                .min(totalEntities, Math.min(numberOfEligibleDataNodes * maxAdBatchTaskPerNode, maxRunningEntitiesPerDetector));
```

Comment:
Copy from https://github.com/opensearch-project/anomaly-detection/pull/58#discussion_r640967506

> How would you handle 10 historical runs at the same time? So if we have 1K entities per detector, 2 data nodes, and maxAdBatchTaskPerNode is 2 and maxRunningEntitiesPerDetector 2. So each historical task run 2 tasks. And 10 historical tasks run 20 tasks. Is my understanding correct?

If maxAdBatchTaskPerNode is 2 and maxRunningEntitiesPerDetector 2, so each HC detector can only run 2 entities, and all tasks running on each data node can't exceed 2, so totally the cluster will have 2 * 2(data nodes) = 4 tasks running.